src/input/type_record.cc

```diff
@@ -359,6 +337,61 @@ Record &Record::has_obligatory_type_key() {
 }
 
 
+TypeBase::MakeInstanceReturnType Record::make_instance(std::vector<ParameterPair> vec) const {
+	Record rec = this->deep_copy();
+	// Replace keys of type Parameter
+	for (std::vector<Key>::iterator key_it=rec.data_->keys.begin(); key_it!=rec.data_->keys.end(); key_it++) {
+		if ( key_it->key_ != "TYPE" ) { // TYPE key isn't substituted
+			MakeInstanceReturnType inst = key_it->type_->make_instance(vec);
+			key_it->type_ = inst.first;
+			rec.add_to_parameter_map(inst.second);
+		}
+	}
+	// Set attributes
+	rec.add_attribute("parameters", rec.print_parameter_map_to_json());
+	std::stringstream type_stream;
+	type_stream << "\"" << this->content_hash() << "\"";
+	rec.add_attribute("generic_type", type_stream.str());
+
+	return std::make_pair( boost::make_shared<Record>(rec.close()), rec.parameter_map_ );
+}
+
+
+Record Record::deep_copy() const {
+	Record rec = Record();
+	rec.data_ =  boost::make_shared<Record::RecordData>(*this->data_);
+	rec.attributes_ = boost::make_shared<attribute_map>(*attributes_);
+	return rec;
+}
+
+
+const Record &Record::add_parent(AbstractRecord &parent) const {
+	ASSERT( parent.is_closed(), "Parent AbstractRecord '%s' must be closed!\n", parent.type_name().c_str());
+
+	// check if parent exists in parent_ptr_ vector
+	if ( data_->parent_ptr_.size() ) {
+		TypeHash hash = parent.content_hash();
+		for (auto it = data_->parent_ptr_.begin(); it != data_->parent_ptr_.end(); ++it) {
+			if ( (*it)->content_hash() == hash ) {
+				return *this;
+			}
+		}
+	}
+
+	data_->parent_ptr_.push_back( boost::make_shared<AbstractRecord>(parent) );
+
+	// finish inheritance
+	ASSERT( data_->keys.size() > 0 && data_->keys[0].key_ == "TYPE",
+				"Derived record '%s' must have defined TYPE key!\n", this->type_name().c_str() );
+	boost::shared_ptr<TypeBase> type_copy = boost::make_shared<Selection>( parent.get_type_selection() );
```

Comment:
This will make trouble in multiple inheritance.  TYPE can not have type of child selection of the parent abstract record since there can be more of them. On the other hand we want to store the actual type in form of index from the selection of particular parent. To make it compatible with the type of the TYPE key, we can declare this key as Interger in the Record, and change implementation in the JSONTo Storage so that the TYPE key is processed in AbstractRecord and skipped in Record. But there are much more places to check, so we postpone all this. Please copy this comment into appropriate issue,